### ISSUE TITLE
dotnet publish task uses directory as project name, not actual project name

### ISSUE THREAD

**Issue by IanKemp** [microsoft/azure-pipelines-tasks#11130]
**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: DotNetCoreCLI@2

## Environment
- Server - Azure Pipelines

## Issue Description

We have a classic/non-YAML build pipeline that has a .NET Core task to publish our artifacts. This task has the "Add project name to publish path" checkbox ticked because the solution being built has multiple projects, and we want to publish all of them:
![image](https://user-images.githubusercontent.com/1726434/63097410-419fb380-bf68-11e9-81a6-6a5357a7d7f6.png)

Previously our solution was structured as so:

```
\Foo.Service\Foo.Service.csproj
\Bar.Service\Bar.Service.csproj
```

and the publish task worked correctly because it placed the build output of the two projects into `$(Build.StagingDirectory)\Foo.Service` and `$(Build.StagingDirectory)\Bar.Service`, which were then zipped to `Foo.Service.zip` and `Bar.Service.zip`. All good.

However we recently refactored the solution so it's structured like this:

```
\Foo\Service\Foo.Service.csproj
\Bar\Service\Bar.Service.csproj
```

and now the publish creates `$(Build.StagingDirectory)\Service` for Foo.Service, which then immediately gets nuked because Bar.Service also ends up in `$(Build.StagingDirectory)\Service`, so I end up with a single `Service.zip` which only contains `Bar.Service`. Not so good.

It appears that instead of using the **project filename minus the extension** when "Add project name to publish path" is checked, this task is using the project **directory**, which is incorrect.

**Comment by IanKemp** [comment]
OK, maybe this is a kind-of feature request since nobody else seems to have this problem. And of course my suggestion to use project name instead of directory name would potentially break the current pipelines of people for whom this is working fine.

So amended suggestion:

* Add a new boolean setting (default false) called `enhancedOutputPathGeneration` to this task. If this new setting is enabled **and** `modifyOutputPath = true`, the `--output` argument passed to `dotnet publish` will be further transformed by this task to include the entire relative pathname of the project file in the output directory, by replacing all directory separators in that pathname with dashes.

For example, in my scenario with both `modifyOutputPath = true` and `enhancedOutputPathGeneration = true`, `\Foo\Service\Foo.Service.csproj` would be output to `$(Build.StagingDirectory)\-Foo-Service-Foo.Service`, and `\Bar\Service\Bar.Service.csproj` to `$(Build.StagingDirectory)\-Bar-Service-Bar.Service`.

* Add a new boolean setting (default false) called `validateNoOverwrite` to this task that if enabled, checks if a subsequent publish will overwrite a previous publish, and if so aborts this task with an appropriate error message.

This will assist people who have encountered the same scenario that I did.

**I am currently working on a PR to add this functionality.**

**Comment by anuragc617** [comment]
@IanKemp yes you are right, the task is taking a directory name as a project name. Your suggestion seems reasonable, we will discuss it and let you know. However, for now, you can use multiple dotNetCore publish tasks one for each service and provide the publish path you want as argument like `--configuration $(BuildConfiguration) --output $(build.artifactstagingdirectory)/Foo-Service` and `--configuration $(BuildConfiguration) --output $(build.artifactstagingdirectory)/Bar-Service` respectively.

**Comment by anuragc617** [comment]
Hi @IanKemp, we don't want to add extra inputs to the task. We will try to fix it without breaking the existing usages. If you want, you can also raise the PR for the same.

**Comment by IanKemp** [comment]
@anuragc617 After further looking into this, I believe the optimal - and simplest - solution would be to use the **full relative path of the project file as the publish location** if "Add project name to publish path" is checked. Under this proposal, `\Foo\Service\Foo.Service.csproj` would be published to `\Foo\Service\Foo.Service` (or `\Foo\Service\Foo.Service.zip` if that option is selected).

If this change was made, the configuration we were previously using for our build pipeline (shown in  screenshot in 1st post) would Just Work™. Our deployment pipeline currently looks for artifacts using the glob `$(System.DefaultWorkingDirectory)/**/Foo.Service.zip`, so there also would be no changes required there.

Depending on how many users have similar pipeline configs, it thus might be possible to introduce this fix without any breakage. "Add project name to publish path" might need to be renamed, though.

**Comment by bishal-pdMSFT** [comment]
> I believe the optimal - and simplest - solution would be to use the full relative path of the project file as the publish location if "Add project name to publish path" is checked.

@IanKemp I believe this suggestion is good. However, as you mentioned, this can be a breaking change. We should ideally do this with a Major version bump. However we do not have any current plan to release a new major version. As we haven't heard this complaint from other folks, I think we can punt this till next release happens.

One possible way the fix can be made available for you for now is via a pre-defined release variable. If this variable is set, the task will choose publish path as suggested above. Does this look like an acceptable compromise to you? Would you be willing to raise this as PR?

@hiyadav does this sound okay to you?

**Comment by IanKemp** [comment]
@bishal-pdMSFT Right now this isn't an issue for my organisation, as we only have 1 affected pipeline and the workaround @anuragc617 provided solves its issue. But I would be interested in providing a PR regardless, as the code it contains can later be repurposed for the next version of this task.

My only questions to you would be regarding the name of the variable (should it be prefixed/suffixed etc.) and if there is a built-in function in your TS code to check if a pipeline variable is defined or not.

**Comment by bishal-pdMSFT** [comment]
@IanKemp sorry for late reply - this went under the radar.
We can have a name like `__PUBLISH_PRESERVE_DIRECTORY_STRUCTURE__`. What do you think?

You can use task lib to get variable value - `tl.getVariable('__PUBLISH_PRESERVE_DIRECTORY_STRUCTURE__')`

**Comment by ttutko** [comment]
Just want to add that we too have run into this issue where 'modifyOutputPath' does not intuitively do what the documentation says or what you would think. Our structure is like so:

Project1\src\project1.csproj
Project2\src\project2.csproj
etc.

This creates a src.zip that clobbers the previous project OR ends up putting all of the files in a single folder regardless of project which is what 'modifyOutputPath' is marketed as fixing but does not.

**Comment by geidans** [comment]
We too have the same issue. Using Blazor WebAssembly template creates folder structures as Client\, Server\, etc. which is fine. However, during the DevOps build it comes Server.zip

**Comment by dglozano** [comment]
We are having the same problem. We have our projects under a /src folder, which causes all projects to become `src.zip`

**Comment by cribeiro84** [comment]
Hey Team
I'm also facing the same problem. Is there any real workaround without having to create multiple tasks once I have 30 projects to be published?

**Comment by menees** [comment]
I worked around it by removing the `--output $(build.artifactstagingdirectory)` argument and letting each project publish into its default relative subfolder (e.g., `Apps/Proj1/bin/Release/net5.0/publish`). Then I added a single PowerShell script step to the pipeline to copy each `publish` subfolder to the `$(build.artifactstagingdirectory)` using the relative paths of my choice. The projects to publish/copy can be passed into the script as an array argument (e.g., `param([string[]] $projects = @('Apps/Proj1', 'Libs/Proj2'))`) It's a little extra work, but at least I'm in control of the process. With Powershell's `Compress-Archive` command, you can even do the zipping in the script if needed.

**Comment by stijnherreman** [comment]
The task documentation itself was fixed a while ago in #12131 to say "project's folder name" instead of "project file name".
Interesting to note that the task does not the path resolve to a `.csproj` file unless you explicitly tell it to do so:
* `src/MyProject` leads to an artifact folder named `src`
* `src/MyProject/` does not work: ##[error]Project file(s) matching the specified pattern were not found.
* `src/MyProject/*.csproj` leads to an artifact folder named `MyProject`

**Comment by patkae** [comment]
This task documentation still says project file name: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/dotnet-core-cli?view=azure-devops#arguments.

Have the changes been published?
